src/server/middleware/repoproxy/proxy.go

```diff
@@ -0,0 +1,95 @@
+//  Copyright Project Harbor Authors
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+package repoproxy
+
+import (
+	"context"
+	"github.com/goharbor/harbor/src/common/models"
+	"github.com/goharbor/harbor/src/common/secret"
+	"github.com/goharbor/harbor/src/common/security"
+	securitySecret "github.com/goharbor/harbor/src/common/security/secret"
+	"github.com/goharbor/harbor/src/controller/project"
+	"github.com/goharbor/harbor/src/lib"
+	"github.com/goharbor/harbor/src/lib/errors"
+	serverHttp "github.com/goharbor/harbor/src/lib/http"
+	"github.com/goharbor/harbor/src/lib/log"
+	"github.com/goharbor/harbor/src/pkg/distribution"
+	"github.com/goharbor/harbor/src/server/middleware"
+	"net/http"
+)
+
+// isProxyProject check the project is a proxy project
+func isProxyProject(p *models.Project) bool {
+	if p == nil {
+		return false
+	}
+	return p.RegistryID > 0
+}
+
+// isProxySession check if current security context is proxy session
+func isProxySession(ctx context.Context) bool {
+	sc, ok := security.FromContext(ctx)
+	if !ok {
+		log.Error("Failed to get security context")
+		return false
+	}
+	if sc.Name() == securitySecret.SecurityContextName &&
+		sc.GetUsername() == secret.ProxyserviceUser {
+		return true
+	}
+	return false
+}
+
+// ManifestPushMiddleware disable push artifact to a proxy project for non proxy session
+func ManifestPushMiddleware() func(http.Handler) http.Handler {
+	return middleware.New(func(w http.ResponseWriter, r *http.Request, next http.Handler) {
+		ctx := r.Context()
+		art := lib.GetArtifactInfo(ctx)
+		p, err := project.Ctl.GetByName(ctx, art.ProjectName)
+		if err != nil {
+			serverHttp.SendError(w, err)
+			return
+		}
+		if isProxyProject(p) && !isProxySession(ctx) {
+			serverHttp.SendError(w,
+				errors.ForbiddenError(
+					errors.Errorf("can not push artifact to a proxy project: %v", p.Name)))
+			return
+		}
+		next.ServeHTTP(w, r)
+		return
+	})
+}
+
+// BlobPushMiddleware disable blob push for non proxy session
+func BlobPushMiddleware() func(http.Handler) http.Handler {
```

Comment:
`ManifestPushMiddleware` and `BlobPushMiddleware`  can be merged into one?

Reply:
+1

Reply:
Merged